Fix segment segfault when join inner plan contains outer params.

bimboterminator1 opened the pull request:
The call of the `cdbpullup_findEclassInTargetList` function takes place when one need to check whether the given `EquivalenceClass` contains exprs that can be expressed in terms of the given `targetlist`. This commit changes the way how equivalence class is iterated through in order to find a eclass member, which is already inside the `targetList` or uses no Vars that are not in `targetList`. Previous code returned the first member it found suitable: if the current member is const or is equal to `targetList` entry, it will be returned immediately, otherwise we check whether the member has non-Var type and is not inside`targetList`, and return the member or continue the iteration process. That logic is optimal in most cases, however, there are the cases when that logic brings erroneous behaviour. For example, in queries containing parameterized subqueries (like the one presented in the tests) it may lead to passing params across motion nodes, which may create the situation when these params are being referenced in different slices, and it probably could lead to segfault when the executor is accessing its attribute list. 

In the query shown in test section occurs when the Flow node's `hashExprs` list is filled during join inner plan creation for scan node. Because subquery has WHERE clause, the EquivalenceClass for subquery's relation distribution key contains two members: one is of type T_Param and other corresponds to distirbution key. During the `hashExprs` list is filled the `cdbpullup_findEclassInTargetList` function is called, and for the given eclass and `targetList` it returns the Param member. This Param is returned because EquivalenceClass member holding it has `em_is_const` flag on, for situation like the one in the tests the flag is set anyway, because WHERE operand refers to outer relations. The function under discussion returns the Param because it came across the appropriate member without checking `targetList`. Eventually this Param entry is appended to `hashExprs` and later to Seq Scan's `targetList` from `hashExprs`, what leads to invalid nested loop plan and possible segfault during execution step.

The proposed solution makes `cdbpullup_findEclassInTargetList` delay the return of consts and params in order to find a member, which is equal to targetList entry. This approach solves the issue with param passage, because for such situations it is guaranteed that plan's targetList contains the distribution key, and if we iterate through eclass and targetList first, the member corresponding to distribution key will be returned from the function instead of the Param. This approach does not break the function's original purpose, because the entires of eclass may come in arbitrary order.

bimboterminator1 committed:
Fix segment segfault when join lateral inner plan contains where clause.

When join lateral subquery contains both LIMIT clause and WHERE clause,
and WHERE clause's operands are refs to outer query relation, for the
inner plan it may put these refs into targetList of the bottom scan node.
This is not correct since it leads to passing params across motion nodes
and it probably could lead to segfault when the executor accessing its
attribute list. A typical case is shown below:

```
create table t1( t1c1 text ) distributed by (t1c1);
create table t2( t2c1 text ) distributed by (t2c1);
explain verbose select * from t2 join lateral
(select * from t1 where t2.t2c1 = t1.t1c1 limit 1) x on true;
                            QUERY PLAN
-----------------------------------------------------------------------
 Nested Loop
   Output: t2.t2c1, t1.t1c1
   ->  Gather Motion 3:1  (slice1; segments: 3)
          Output: t2.t2c1
         ->  Seq Scan on public.t2
               Output: t2.t2c1
   ->  Materialize
          Output: t1.t1c1
         ->  Limit
                Output: t1.t1c1
               ->  Result
                      Output: t1.t1c1
                     Filter: ((t2.t2c1) = t1.t1c1)
                     ->  Materialize
                            Output: t1.t1c1, (t2.t2c1)
                           ->  Gather Motion 3:1  (slice2; segments: 3)
                                  Output: t1.t1c1, (t2.t2c1)
                                 ->  Seq Scan on public.t1
                                        Output: t1.t1c1, t2.t2c1

```
The above plan is invalid because Seq Scan on t1 is prefomed at slice2
and uses params that are scanned from t2 at slice1. And the presence of
such params may lead to segfault during inner plan execution.

This commit fixes the bug by handling the targetList completion procedure,
which takes place during plan creation for the subquery. The root of the
problem is in appending extra entries of type T_Param, which refer to outer
relations, to final targetList from hashExprs list in plan's Flow structure.
The hashExprs list is filled with distribution key and, in this case,
equivalence class member, which appears with the WHERE clause. The issue is
that during plan construction it is hard to predict and control the slices,
where the params will be evaluated. Therefore, in order to exclude such phenomena
the proposed solution filters these extra params and prevents subplan
from holding outer refs in similar cases.

RekGRpth commented on src/test/regress/sql/join_gp.sql:
```diff
@@ -712,3 +712,26 @@ where x.a + 1 in (select b from t2_dedupsemi_indexonly);
+insert into t1_lateral_where values (1), (2), (3);
+insert into t2_lateral_where values (3), (4);
+
+explain verbose select * from t2_lateral_where as t2 inner join lateral
```
Maybe use `explain (verbose, costs off)`?

bimboterminator1 replied:
Yes, it would be better without costs.

Stolb27 commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -667,12 +667,27 @@ disuse_physical_tlist(PlannerInfo *root, Plan *plan, Path *path)
+				{
+					Expr	*expr = (Expr *) lfirst(lc);
+
+					if (!IsA(expr, Param))
```
> The hashExprs list consists of either distribution key or it's EquivalenceClass members, which could appear with WHERE clause.

hashExpr of Flow [must contain](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/cdb/cdbpathlocus.c#L365-L377) only distribution keys references for scan nodes, represented as references to [equivalence class](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/README#L397) lists. As far I understand, equivalence classes are built not only for where clauses, but for join condition too. Do we have guaranties of filtering other relation Vars?

```c
 *      The partitioning key is represented by a List of DistributionKeys,
 *      one for each key column. Each DistributionKey contains a list of
 *      EquivalenceClasses, which contain expressions that can be used
 *      to compute the value for the key column. Any of the expressions
 *      can be used to compute the value, depending on what relations are
 *      available at that part of the plan.
 *
 *      For example, if the query contains a "WHERE a=b" clause, the planner
 *      would form an EquivalenceClass that contains two members, "a" and "b".
 *      Because of the WHERE clause, either "a" and "b" can be used to
 *      compute the hash value. Usually, a DistributionKey contains only one
 *      EquivalenceClass, because whenever there is an equijoin on two
 *      expressions, the planner puts them in the same EquivalenceClass.
```

Also, please describe in the PR description why we must add this fields to TL.

bimboterminator1 replied:
The entries may have type T_Param only when we are in a subquery, on one hand, it indicates that the param has come down from outer scope.  And in any other case it is possible, for example during joins, that T_Var in hashExprs will be from other relation. Howerver, in `disuse_physical_tlist`, before calling `add_to_flat_tlist_junk`, there is a call of [build_path_tlist](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L667), which makes the targetList contain only entries actually referenced (this also includes operand from where, join restriction). And when appending equivalent var entity of other relation from hashExprs to targetList, there is a [check](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/util/tlist.c#L199) in `add_to_flat_tlist_junk` , which does not allow to add expression that is already in targetList. And when we have where clause or join conditions,   there could be already the var from current relation, not other. And the [comparison](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/nodes/equalfuncs.c#L165-L175) function is this case will return true, because location attribute is [unchecked](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/nodes/equalfuncs.c#L104-L105). Although var entry  refers to other relation, it will not be added to targetList, or, if somehow there is no equivelent member in targetList, this var will be enentually added as current relation var (not other relation). Again, the problem occurs only when we have LIMIT, it leads to nested loop in different slices, other cases are really likely to be safe. And I updated PR description, I added 2 extra sentences stating that there is a need in passing a distribution key.

Stolb27 replied:
I've recheck existing solution for query from tests with debugger:

```
Thread 1 "postgres" hit Breakpoint 3, disuse_physical_tlist (root=0x55e5b9b38308, plan=0x55e5b9adfa88, path=0x55e5b9a00b30) at createplan.c:667
(gdb) n
(gdb) p nodeToString(plan->flow->hashExprs)
$2 = 0x55e5b9ae0d18 "({PARAM :paramkind 1 :paramid 0 :paramtype 25 :paramtypmod -1 :paramcollid 100 :location 83})"
```

`hashExprs` contains only PARAM entry, not whole Equivalence class. Am I correct? In this case, I think that it's an invalid hash entry value, and we must research the reason of building such value. As far I understand, new solutions tries to fix this problem.

Stolb27 commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -667,12 +667,27 @@ disuse_physical_tlist(PlannerInfo *root, Plan *plan, Path *path)
 			 */
 			if (plan->flow && plan->flow->hashExprs)
 			{
-				plan->targetlist = add_to_flat_tlist_junk(plan->targetlist, plan->flow->hashExprs, true /* resjunk */);
```
You've replaced `add_to_flat_tlist_junk` function with a custom logic that filter Params. But there are several places of usage `add_to_flat_tlist_junk` in the gpdb codebase that also modify target list. Why should we patch only this place?

bimboterminator1 replied:
`add_to_flat_tlist_junk` with `hashExprs` is called in 2 more places: 
**1.** Inside `cdbpathtoplan_create_motion_plan`. This function is called from `create_motion_plan` after the call of `disuse_physical_tlist` for [subpath](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L1396-L1399) of CdbMotionPath.  The `add_to_flat_tlist_junk` is called for hashExprs in case of [redistribute motion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7217-L7251) . Consider example for hash join, we have outer_plan created with redistibute motion:
```
create table t2( t2c1 int, a int) distributed by (t2c1);
create table t1( t1c1 int, a int ) distributed by (t1c1);
explain verbose select * from t2 join lateral
(select * from t1 where t2.t2c1 = t1.a ) x on true;
                                           QUERY PLAN                                              
-----------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  
   Output: t2.t2c1, t2.a, t1.t1c1, t1.a
   ->  Hash Join 
         Output: t2.t2c1, t2.a, t1.t1c1, t1.a
         Hash Cond: (t1.a = t2.t2c1)
         ->  Redistribute Motion 3:3  (slice1; segments: 3) 
               Output: t1.t1c1, t1.a
               Hash Key: t1.a
               ->  Seq Scan on public.t1 
                     Output: t1.t1c1, t1.a
         ->  Hash 
               Output: t2.t2c1, t2.a
               ->  Seq Scan on public.t2
                     Output: t2.t2c1, t2.a
 Optimizer: Postgres query optimizer
(15 rows)
```
When we are creating the outer plan for hashjoin, `create_motion_plan` is called, where for SeqScan part `disuse_physical_tlist` is called too. Then we fall into `cdbpathtoplan_create_motion_plan`  [redistribute motion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7217-L7251) case, where we try to add into `subplan->targetList` (subplan here corresponds to the SeqSqan on t1) `hashExprs` taken for motion locus. `hashExprs` here contain the keys on which the redistribition will perform:
```
hashExprs (
      {VAR 
      :varno 4 
      :varattno 2       <----- t1.a 
      :vartype 23 
      :vartypmod -1 
      :varcollid 0 
      :varlevelsup 0 
      :varnoold 4 
      :varoattno 2 
      :location 80
      }
   )
``` 
t1.a is being added in hashExprs in `cdbpathlocus_get_distkey_exprs` func, it iterates through equivalence class members and do several checks whether member is already in targetList ( in `cdbpullup_findEclassInTargetList`). And in this case it is already inside the targetList because of previous `disuse_physical_tlist` call, which adds t1.a to SeqScan TL, because it is in select list (if not it would be added because of WHERE clause with t1.a as an operand).

Next,  `add_to_flat_tlist_junk` will try to add to SeqScan's targetList t1.a. And again there is a check on presenece inside the TL. It presents.
Why the filtration of params is unneccesary in that case? For most of the joins, excepting nested loop, the lateral refernce will be added to all EquivalenceClasses as Var, not Param, because the lateral subquery is not really considered as a separate subquery here. Once more, I think the main problem in limit clause causing nested loop processing.  And It would be better filter out params only in that case, because I am afraid that params could occur in another cases and there would be optimization losses. But I still can't imagine and can't find another example query with param inside EquivalenceClass. In other words, I can't proof my solution strongly.
Probably Andrey Sokolov's solution is better because it prevents from adding anything into hashExprs. And all `add_to_flat_tlist_junk` will be safe in that case. 

**2.** Inside grouping_planner function the `add_to_flat_tlist_junk` is called for [window function plan's tlist.](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/planner.c#L2581-L2589) . And in this case there is no guaranties of external Params absence. Again, probably Andrey's Approach will make it safer

Stolb27 replied:
> Then we fall into cdbpathtoplan_create_motion_plan [redistribute motion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7217-L7251) case, where we try to add into subplan->targetList (subplan here corresponds to the SeqSqan on t1) hashExprs taken for motion locus.

only in case of `hashExprs` [contains](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7234) a sub-plan reference. Your example plan doesn't contain subplans.

bimboterminator1 committed:
Revert commit with filtering solution

bimboterminator1 committed:
Fix segment segfault when join inner plan contains outer params.

When a subquery has WHERE clause with equality condition between the
relation's distribution key and params, which refer to the relations from
upper query levels (for example, lateral reference in join subquery), then
the distiribution key's EquivalenceClass may contain these params. Therefore,
these params may be put into the subplan node's targetList. There are the
cases when such behaviour leads to these params are being referenced in
different slices, and it probably could lead to segfault when the executor
is accessing its attribute list. A typical case is shown below:

```
create table t1( t1c1 text ) distributed by (t1c1);
create table t2( t2c1 text ) distributed by (t2c1);
explain verbose select * from t2 join lateral
(select * from t1 where t2.t2c1 = t1.t1c1 limit 1) x on true;
                            QUERY PLAN
-----------------------------------------------------------------------
 Nested Loop
   Output: t2.t2c1, t1.t1c1
   ->  Gather Motion 3:1  (slice1; segments: 3)
          Output: t2.t2c1
         ->  Seq Scan on public.t2
               Output: t2.t2c1
   ->  Materialize
          Output: t1.t1c1
         ->  Limit
                Output: t1.t1c1
               ->  Result
                      Output: t1.t1c1
                     Filter: ((t2.t2c1) = t1.t1c1)
                     ->  Materialize
                            Output: t1.t1c1, (t2.t2c1)
                           ->  Gather Motion 3:1  (slice2; segments: 3)
                                  Output: t1.t1c1, (t2.t2c1)
                                 ->  Seq Scan on public.t1
                                        Output: t1.t1c1, t2.t2c1

```
The above plan is invalid because Seq Scan on t1 is performed at slice2
and uses params that are scanned from t2 at slice1. And the presence of
such params may lead to segfault during inner plan execution.

This commit fixes the bug by changing the cdbpullup_findEclassInTargetList
function without breaking its inherent behaviour and purpose. The call of the
function takes place when the search through the given equivalence class
for a member, which is already inside the targetList or uses no Vars that are
not in targetList, is required. In query above it is used to fill the Flow's
hashExprs list during plan creation for the bottom Seq Scan. The root of the
problem in case above is that hashExprs contains entry of type T_Param,
which refer to outer relations, and this param is appended to final targetList
from the hashExprs. In this case the hashExprs list is filled with the equivalence
class member of type T_Param, which appears with the WHERE clause. The proposed
solution makes the cdbpullup_findEclassInTargetList function iterate through whole
equivalence class and through whole targetList first, hoping to find member, which
is equal to targetList entry. And only next step is to check whether
member is pseudoconst or non-Var that is not in targetList, because for
given query Param is added due to these conditions. These changes prevent subplan
from holding outer refs in similar cases.

bimboterminator1 committed:
Revert commit with loops separation solution

bimboterminator1 committed:
Fix segment segfault when join inner plan contains outer params.

The call of the cdbpullup_findEclassInTargetList function takes place when
one need to check whether the given EquivalenceClass contains exprs that
can be expressed in terms of the given targetlist. This commit changes the way
how equivalence class is iterated through in order to find a eclass member,
which is already inside the targetList or uses no Vars that are not in
targetList. Previous code returned the first member it found suitable: if the
current member is const or is equal to targetList entry, it will be returned
immediately, otherwise we check whether the member has non-Var type and is not
inside targetList, and return the member or continue the iteration process. That
logic is optimal in most cases, however, there are the cases when that logic
brings erroneous behaviour. For example, in queries containing parameterized
subqueries (like the one presented in the tests) it may lead to passing params
across motion nodes, which may create the situation when these params are being
referenced in different slices, and it probably could lead to segfault when the
executor is accessing its attribute list.

In the query shown in test section such situtation occurs when the Flow node's
hashExprs list is filled during join inner plan creation for scan node. Because
subquery has WHERE clause, the EquivalenceClass for subquery's relation
distribution key contains two members: one is of type T_Param and other
corresponds to distirbution key. When hashExprs list is being filled the
cdbpullup_findEclassInTargetList function is called, and for the given eclass
and targetList it returns the Param member. This Param is returned because
EquivalenceClass member holding it has em_is_const flag on, for situation like
the one in the tests the flag is set anyway, because WHERE operand refers to
outer relations. The function under discussion returns the Param because it
came across the appropriate member without checking targetList. Eventually
this Param entry is appended to hashExprs and later to Seq Scan's targetList
from hashExprs, what leads to invalid nested loop plan and possible segfault
during execution step.

The proposed solution makes cdbpullup_findEclassInTargetList delay the return
of consts and params in order to find a member, which is equal to targetList
entry. This approach solves the issue with param passage, because for such
situations it is guaranteed that plan's targetList contains the distribution
key, and if we iterate through eclass and targetList first, the member
corresponding to distribution key will be returned from the function instead
of the Param. This approach does not break the function's original purpose,
because the entires of eclass may come in arbitrary order.

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3331
